service/frontend/workflow_handler.go

```diff
@@ -2817,6 +2821,62 @@ func (wh *WorkflowHandler) DescribeSchedule(ctx context.Context, request *workfl
 	}, nil
 }
 
+func (wh *WorkflowHandler) annotateSearchAttributesOfScheduledWorkflow(
+	queryResponse *schedspb.DescribeResponse,
+	nsName string,
+) error {
+	ei, err := wh.getScheduledWorkflowExecutionInfoFrom(queryResponse)
+	if err != nil {
+		return err
+	}
+	annotatedAttributes, err := wh.createAnnotatedSearchAttributesFromNewWorkflow(ei, nsName)
+	if err != nil {
+		return fmt.Errorf("annotate search attributes:%w", err)
+	}
+	ei.SearchAttributes = annotatedAttributes
+	return nil
+}
+
+func (wh *WorkflowHandler) getScheduledWorkflowExecutionInfoFrom(
+	queryResponse *schedspb.DescribeResponse,
+) (*workflowpb.NewWorkflowExecutionInfo, error) {
+	action := queryResponse.GetSchedule().GetAction().GetAction()
+	startWorkflowAction, ok := action.(*schedpb.ScheduleAction_StartWorkflow)
+	if !ok {
+		return nil, serviceerror.NewInvalidArgument("unsupported action kind")
+	}
+	newWorkflowExecutionInfo := startWorkflowAction.StartWorkflow
+	return newWorkflowExecutionInfo, nil
+}
+
+func (wh *WorkflowHandler) createAnnotatedSearchAttributesFromNewWorkflow(
```

Comment:
It feels like this "annotating search attributes" is conceptually a filter on search attributes, i.e. it should accept `*commonpb.SearchAttributes` and return the same thing. Pulling the search attributes out of `NewWorkflowExecutionInfo` could be in the caller.

Additionally, in that case it makes it easy to put a `if searchAttrs == nil { return nil }` at the top and avoid a bunch of work in that very common case.